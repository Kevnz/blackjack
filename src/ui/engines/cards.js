export function getName(value) {
  var name
  switch (value) {
    case 1:
      name = 'ace'
      break
    case 11:
      name = 'jack'
      break
    case 12:
      name = 'queen'
      break
    case 13:
      name = 'king'
      break
    default:
      name = value
  }
  return name
}

export function shuffleArray(array) {
  for (var i = array.length - 1; i > 0; i--) {
    var j = Math.floor(Math.random() * (i + 1))
    var temp = array[i]
    array[i] = array[j]
    array[j] = temp
  }
  return array
}

export function buildSuit(suit) {
  var cards = []

  for (var i = 1; i <= 13; i++) {
    var cardname = getName(i)
    var cardIMAGE = '/assets/card' + suit + cardname + '.png'
    if (i === 1 || i > 10) {
      cardIMAGE = '/assets/card' + suit + cardname.slice(0, 1) + '.png'
    }

    cards.push({ val: i, suit: suit, name: cardname, img: cardIMAGE })
  }

  return cards
}
export function buildDeck() {
  const spades = shuffleArray(buildSuit('spades'))
  const diamonds = shuffleArray(buildSuit('diamonds'))
  const hearts = shuffleArray(buildSuit('hearts'))
  const clubs = shuffleArray(buildSuit('clubs'))

  return shuffleArray(spades.concat(diamonds, hearts, clubs))
}
