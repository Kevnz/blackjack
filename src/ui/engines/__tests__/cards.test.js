import { buildSuit, buildDeck } from '../cards'

describe('The base card engine', () => {
  it('should build a suit of cards', () => {
    const hearts = buildSuit('hearts')
    const clubs = buildSuit('clubs')
    const spades = buildSuit('spades')
    const diamonds = buildSuit('diamonds')

    expect(hearts).toMatchSnapshot()
    expect(clubs).toMatchSnapshot()
    expect(spades).toMatchSnapshot()
    expect(diamonds).toMatchSnapshot()
  })
  it('should build a deck of 52 cards with 4 suits and ace through king', () => {
    const deck = buildDeck()
    // deck
    expect(deck.length).toBe(52)
    const diamonds = deck.filter(c => c.suit === 'diamonds')
    expect(diamonds.length).toBe(13)
    expect(diamonds.find(c => c.name === 'ace')).toBeDefined()
    expect(diamonds.find(c => c.name === 'jack')).toBeDefined()
    expect(diamonds.find(c => c.name === 'king')).toBeDefined()
    expect(diamonds.find(c => c.name === 'queen')).toBeDefined()
    expect(diamonds.find(c => c.name === 2)).toBeDefined()
    expect(diamonds.find(c => c.name === 3)).toBeDefined()
    expect(diamonds.find(c => c.name === 4)).toBeDefined()
    expect(diamonds.find(c => c.name === 5)).toBeDefined()
    expect(diamonds.find(c => c.name === 6)).toBeDefined()
    expect(diamonds.find(c => c.name === 7)).toBeDefined()
    expect(diamonds.find(c => c.name === 8)).toBeDefined()
    expect(diamonds.find(c => c.name === 9)).toBeDefined()
    expect(diamonds.find(c => c.name === 10)).toBeDefined()

    const hearts = deck.filter(c => c.suit === 'hearts')
    expect(hearts.length).toBe(13)
    const spades = deck.filter(c => c.suit === 'spades')
    expect(spades.length).toBe(13)
    const clubs = deck.filter(c => c.suit === 'clubs')
    expect(clubs.length).toBe(13)
  })
})
