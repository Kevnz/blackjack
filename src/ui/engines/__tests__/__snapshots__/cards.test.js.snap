// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The base card engine should build a suit of cards 1`] = `
Array [
  Object {
    "img": "/assets/cardheartsa.png",
    "name": "ace",
    "suit": "hearts",
    "val": 1,
  },
  Object {
    "img": "/assets/cardhearts2.png",
    "name": 2,
    "suit": "hearts",
    "val": 2,
  },
  Object {
    "img": "/assets/cardhearts3.png",
    "name": 3,
    "suit": "hearts",
    "val": 3,
  },
  Object {
    "img": "/assets/cardhearts4.png",
    "name": 4,
    "suit": "hearts",
    "val": 4,
  },
  Object {
    "img": "/assets/cardhearts5.png",
    "name": 5,
    "suit": "hearts",
    "val": 5,
  },
  Object {
    "img": "/assets/cardhearts6.png",
    "name": 6,
    "suit": "hearts",
    "val": 6,
  },
  Object {
    "img": "/assets/cardhearts7.png",
    "name": 7,
    "suit": "hearts",
    "val": 7,
  },
  Object {
    "img": "/assets/cardhearts8.png",
    "name": 8,
    "suit": "hearts",
    "val": 8,
  },
  Object {
    "img": "/assets/cardhearts9.png",
    "name": 9,
    "suit": "hearts",
    "val": 9,
  },
  Object {
    "img": "/assets/cardhearts10.png",
    "name": 10,
    "suit": "hearts",
    "val": 10,
  },
  Object {
    "img": "/assets/cardheartsj.png",
    "name": "jack",
    "suit": "hearts",
    "val": 11,
  },
  Object {
    "img": "/assets/cardheartsq.png",
    "name": "queen",
    "suit": "hearts",
    "val": 12,
  },
  Object {
    "img": "/assets/cardheartsk.png",
    "name": "king",
    "suit": "hearts",
    "val": 13,
  },
]
`;

exports[`The base card engine should build a suit of cards 2`] = `
Array [
  Object {
    "img": "/assets/cardclubsa.png",
    "name": "ace",
    "suit": "clubs",
    "val": 1,
  },
  Object {
    "img": "/assets/cardclubs2.png",
    "name": 2,
    "suit": "clubs",
    "val": 2,
  },
  Object {
    "img": "/assets/cardclubs3.png",
    "name": 3,
    "suit": "clubs",
    "val": 3,
  },
  Object {
    "img": "/assets/cardclubs4.png",
    "name": 4,
    "suit": "clubs",
    "val": 4,
  },
  Object {
    "img": "/assets/cardclubs5.png",
    "name": 5,
    "suit": "clubs",
    "val": 5,
  },
  Object {
    "img": "/assets/cardclubs6.png",
    "name": 6,
    "suit": "clubs",
    "val": 6,
  },
  Object {
    "img": "/assets/cardclubs7.png",
    "name": 7,
    "suit": "clubs",
    "val": 7,
  },
  Object {
    "img": "/assets/cardclubs8.png",
    "name": 8,
    "suit": "clubs",
    "val": 8,
  },
  Object {
    "img": "/assets/cardclubs9.png",
    "name": 9,
    "suit": "clubs",
    "val": 9,
  },
  Object {
    "img": "/assets/cardclubs10.png",
    "name": 10,
    "suit": "clubs",
    "val": 10,
  },
  Object {
    "img": "/assets/cardclubsj.png",
    "name": "jack",
    "suit": "clubs",
    "val": 11,
  },
  Object {
    "img": "/assets/cardclubsq.png",
    "name": "queen",
    "suit": "clubs",
    "val": 12,
  },
  Object {
    "img": "/assets/cardclubsk.png",
    "name": "king",
    "suit": "clubs",
    "val": 13,
  },
]
`;

exports[`The base card engine should build a suit of cards 3`] = `
Array [
  Object {
    "img": "/assets/cardspadesa.png",
    "name": "ace",
    "suit": "spades",
    "val": 1,
  },
  Object {
    "img": "/assets/cardspades2.png",
    "name": 2,
    "suit": "spades",
    "val": 2,
  },
  Object {
    "img": "/assets/cardspades3.png",
    "name": 3,
    "suit": "spades",
    "val": 3,
  },
  Object {
    "img": "/assets/cardspades4.png",
    "name": 4,
    "suit": "spades",
    "val": 4,
  },
  Object {
    "img": "/assets/cardspades5.png",
    "name": 5,
    "suit": "spades",
    "val": 5,
  },
  Object {
    "img": "/assets/cardspades6.png",
    "name": 6,
    "suit": "spades",
    "val": 6,
  },
  Object {
    "img": "/assets/cardspades7.png",
    "name": 7,
    "suit": "spades",
    "val": 7,
  },
  Object {
    "img": "/assets/cardspades8.png",
    "name": 8,
    "suit": "spades",
    "val": 8,
  },
  Object {
    "img": "/assets/cardspades9.png",
    "name": 9,
    "suit": "spades",
    "val": 9,
  },
  Object {
    "img": "/assets/cardspades10.png",
    "name": 10,
    "suit": "spades",
    "val": 10,
  },
  Object {
    "img": "/assets/cardspadesj.png",
    "name": "jack",
    "suit": "spades",
    "val": 11,
  },
  Object {
    "img": "/assets/cardspadesq.png",
    "name": "queen",
    "suit": "spades",
    "val": 12,
  },
  Object {
    "img": "/assets/cardspadesk.png",
    "name": "king",
    "suit": "spades",
    "val": 13,
  },
]
`;

exports[`The base card engine should build a suit of cards 4`] = `
Array [
  Object {
    "img": "/assets/carddiamondsa.png",
    "name": "ace",
    "suit": "diamonds",
    "val": 1,
  },
  Object {
    "img": "/assets/carddiamonds2.png",
    "name": 2,
    "suit": "diamonds",
    "val": 2,
  },
  Object {
    "img": "/assets/carddiamonds3.png",
    "name": 3,
    "suit": "diamonds",
    "val": 3,
  },
  Object {
    "img": "/assets/carddiamonds4.png",
    "name": 4,
    "suit": "diamonds",
    "val": 4,
  },
  Object {
    "img": "/assets/carddiamonds5.png",
    "name": 5,
    "suit": "diamonds",
    "val": 5,
  },
  Object {
    "img": "/assets/carddiamonds6.png",
    "name": 6,
    "suit": "diamonds",
    "val": 6,
  },
  Object {
    "img": "/assets/carddiamonds7.png",
    "name": 7,
    "suit": "diamonds",
    "val": 7,
  },
  Object {
    "img": "/assets/carddiamonds8.png",
    "name": 8,
    "suit": "diamonds",
    "val": 8,
  },
  Object {
    "img": "/assets/carddiamonds9.png",
    "name": 9,
    "suit": "diamonds",
    "val": 9,
  },
  Object {
    "img": "/assets/carddiamonds10.png",
    "name": 10,
    "suit": "diamonds",
    "val": 10,
  },
  Object {
    "img": "/assets/carddiamondsj.png",
    "name": "jack",
    "suit": "diamonds",
    "val": 11,
  },
  Object {
    "img": "/assets/carddiamondsq.png",
    "name": "queen",
    "suit": "diamonds",
    "val": 12,
  },
  Object {
    "img": "/assets/carddiamondsk.png",
    "name": "king",
    "suit": "diamonds",
    "val": 13,
  },
]
`;
